rust/parser/src/simulator.rs

```diff
@@ -0,0 +1,126 @@
+//! Simulates parser execution, for a single token of input, without incurring
+//! any side effects.
+//!
+//! This is basically a copy of the parser.rs source code with calls to
+//! generated_parser::reduce, and stack bookkeeping, omitted.
+
+use ast::SourceLocation;
+use generated_parser::{
+    ErrorCode, Result, TerminalId, Token, TABLES,
+};
+use crate::parser::{Action, Parser};
+
+pub struct Simulator<'parser> {
+    sp: usize,
+    state: usize,
+    state_stack: &'parser [usize],
+}
+
+impl<'parser> Simulator<'parser> {
+    pub fn new(state_stack: &'parser [usize]) -> Simulator {
+        let sp = state_stack.len() - 1;
+        Simulator {
+            sp,
+            state: state_stack[sp],
+            state_stack,
+        }
+    }
+
+    fn action(&self, t: TerminalId) -> Action {
+        Parser::action_at_state(t, self.state)
+    }
+
+    // Simulate the action of Parser::reduce_all without calling any AstBuilder
+    // methods or modifying the Parser state. Naturally, "early errors"
+    // detected by AstBuilder methods are not caught.
+    fn reduce_all(&mut self, t: TerminalId) -> Action {
+        let mut action = self.action(t);
+        while action.is_reduce() {
+            let prod_index = action.reduce_prod_index();
+            let (num_pops, nt) = TABLES.reduce_simulator[prod_index];
+            debug_assert!((nt as usize) < TABLES.goto_width);
+            self.sp -= num_pops;
+            let prev_state = self.state_stack[self.sp];
+            let state_after =
+                TABLES.goto_table[prev_state * TABLES.goto_width + nt as usize] as usize;
+            debug_assert!(state_after < TABLES.state_count);
+            self.sp += 1;
+            self.state = state_after;
+            action = self.action(t);
+        }
+        action
+    }
```

Comment:
it would be nice if this can be implemented with something like C++ template method, sharing single code between parser and simulator (so, with single code for the entire logic, and the difference between them are resolved at compile-time),
given copy-paste often result in unexpected diverge and troublesome bug.
same for other methods.
(not necessary to be fixed in this PR tho)